### ISSUE TITLE
4.6 and 4.7.2 output difference (java)

### ISSUE THREAD

**Issue by KakCAT** [antlr/antlr4#2594]
I have the following grammar:

```

grammar Norua;


staticClass:    STATIC_CLASS vID=IDENTIFIER PUNT_COMA ( declExternalFunc )*;

type        : INT ;


declExternalFunc:
                EXTERNAL vType=type vID=IDENTIFIER PAR_OPEN PAR_CLOSE PUNT_COMA            
                ;
   




// discard spaces
WS:   (' ' | '\r' | '\n' | '\t') -> channel(HIDDEN);


// keywords
EXTERNAL: 'external';
STATIC_CLASS: 'static_class';
INT: 'int';


// IDENTIFIERs

IDENTIFIER: Char+;



// basic tokens
PUNT_COMA: ';';
COMA: ',';
PAR_OPEN: '(';
PAR_CLOSE: ')';




fragment Char: ('A'..'Z' | 'a'..'z');
fragment Digit: ('0'..'9');



```

(I've simplified as much as I've been able, I'm not an ANTLR expert)

And the following input to the grammar:

```

static_class Debug;

internal int WriteLine ();

```

(note that **internal** is not a keyword, it should be 'external', thus giving an error)

I compile it with antlr-4.6-complete.jar , then I run "  grun Norua staticClass  -trace < test.rps "

> 
> Warning: TestRig moved to org.antlr.v4.gui.TestRig; calling automatically
> enter   staticClass, LT(1)=static_class
> consume [@0,0:11='static_class',<3>,1:0] rule staticClass
> consume [@2,13:17='Debug',<5>,1:13] rule staticClass
> consume [@3,18:18=';',<6>,1:18] rule staticClass
> line 4:0 mismatched input 'internal' expecting {<EOF>, 'external'}
> exit    staticClass, LT(1)=<EOF>

I get a mismatched input 'internal', as expected.

However, when compiling it with antlr-4.7.2-complete.jar, the output differs:

> 
> Warning: TestRig moved to org.antlr.v4.gui.TestRig; calling automatically
> enter   staticClass, LT(1)=static_class
> consume [@0,0:11='static_class',<3>,1:0] rule staticClass
> consume [@2,13:17='Debug',<5>,1:13] rule staticClass
> consume [@3,18:18=';',<6>,1:18] rule staticClass
> exit    staticClass, LT(1)=internal

As I get no error "mismatched input internal", my generated code can't handle it (i.e. giving an error message) and it messes my whole compilation. (actually I'm using C#, but moved the issue to java to check if it was just a C# port problem)

I've posted this in the news, at first I thought it could be an error on my side, but after some consideration, considering the grammar definition rules have not changed between 4.6 and 4.7.2 (that I've been able to find) it may be a bug 

Thanks!

**Comment by ericvergnaud** [comment]
your grammar only accepts ‘external’ as token after staticClass rule.
close this and get support from the google discussion group

**Comment by KakCAT** [comment]
I know that, the problem is that 4.6 gives an error (line 4:0 mismatched input 'internal' expecting {, 'external'} ) but 4.7.2 gives no error and accepts the input file as correct.

I think this qualifies as a bug, although I'm no expert.  (please confirm me if I should close the issue)

**Comment by KakCAT** [comment]
just to ilustrate the problem. Same input file :
![antlr4_parse_tree_4_6](https://user-images.githubusercontent.com/2372854/61146540-961fc100-a4da-11e9-9670-cc7c4a847971.png)
![antlr4_parse_tree_4_7_2](https://user-images.githubusercontent.com/2372854/61146543-97e98480-a4da-11e9-8a6f-a1038a1a7f04.png)

First image is 4.6, second image is 4.7.2.
It seems like it just stops parsing after finding "internal" and returns no errors on 4.7.2. I was expecting to get the same result as 4.6

**Comment by ericvergnaud** [comment]
since your start rule does not expect an EOF, it is matched. The next token is made available for another invocation using a different start rule.
I believe raising an error in 4.6 was a bug.

**Comment by KakCAT** [comment]
Thanks for your answer. Definitively a bug in 4.6 fixed in 4.7.2, not the contrary  (to be honest I didn't even know EOF existed). Closing
